flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java

```diff
@@ -609,7 +611,7 @@ public void close() throws Exception {
 							startTimestamp);
 
 					//check if newly created state is optional (have a PROCEED path to Final state)
-					final State<T> finalState = findFinalStateAfterProceed(sharedBuffer, nextState, event.getEvent(), computationState);
+					final State<T> finalState = findFinalStateAfterProceed(new ConditionContext<>(this, sharedBuffer, computationState), nextState, event.getEvent());
```

Comment:
Why don't you use the one created in the beginning?

Reply:
I think the sharedbuffer has been changed during the `TAKE` branch, the conditionContext should be different.

Reply:
I think over it again, the content of the `sharedBuffer` make difference  to  the result of the `getEventsForPattern`, so the result should be update with the change of the `sharedBuffer`. But i think we only have to reset the `shouldUpdate` flag to `true` here rather than create a  context again, right? @dawidwys

Reply:
Yes that is true the shared buffer might change but only "after" the CS. I mean those changes can't affect partial match for current Computation State. Therefore I would suggest to remove the flag completely and just keep the lazy evaluation of `matchedEvents` in `ConditionContext` by comparing it to null:

	if (matchedEvents == null) {
		this.matchedEvents = nfa.extractCurrentMatches(sharedBuffer, computationState);
	}

Reply:
Yes, you are right. Read the code again,  the `TAKE` branch only put the new `Node` to sharedBuffer which just point to the previousNodeId, This indeed don't affect the result of the current CS's partial match. I will take your suggestion